File: golang/_0002_Add_Two_Numbers/main.go
package main

import "leetcode/golang/common"

type ListNode = common.ListNode

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	if l1 == nil {
		return l2
	} else if l2 == nil {
		return l1
	} else {
		dummy := &ListNode{}
		cur := dummy
		carry := 0

		for ; l1 != nil || l2 != nil || carry > 0; {
			t := carry

			if l1 != nil {
				t += l1.Val
				l1 = l1.Next
			}

			if l2 != nil {
				t += l2.Val
				l2 = l2.Next
			}

			carry = t / 10

			cur.Next = &ListNode{Val: t % 10}
			cur = cur.Next
		}

		return dummy.Next
	}
}

File: python3/_0002_Add_Two_Numbers/main.py
from python3.common.define import ListNode


class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        dummy = tail = ListNode(0)
        c = 0
        while l1 or l2 or c:
            if l1:
                c += l1.val
                l1 = l1.next

            if l2:
                c += l2.val
                l2 = l2.next

            tail.next = ListNode(c % 10)
            tail = tail.next

            c //= 10

        return dummy.next
